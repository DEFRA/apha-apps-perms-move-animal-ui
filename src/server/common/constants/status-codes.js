/**
 * @typedef {Record<string, number>} StatusCodes
 */
export const statusCodes = {
  ok: 200,
  noContent: 204,
  badRequest: 400,
  unauthorized: 401,
  forbidden: 403,
  notFound: 404,
  imATeapot: 418,
  redirect: 302,
  serverError: 500,
  badGateway: 502,
  gatewayTimeout: 504
}
