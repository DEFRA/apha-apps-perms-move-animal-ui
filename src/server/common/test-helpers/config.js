import { config } from '~/src/config/config.js'

/* global jest */

const configGet = config.get.bind(config)

/**
 * @param {string} key
 * @param {any} value
 */
export const spyOnConfig = (key, value) => {
  jest.spyOn(config, 'get').mockImplementation((name) => {
    if (name === key) {
      return value
    } else {
      return configGet(name)
    }
  })
  return config
}
