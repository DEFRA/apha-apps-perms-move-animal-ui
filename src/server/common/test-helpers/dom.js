import { JSDOM } from 'jsdom'

/**
 * @param {string} payload
 * @returns {Document}
 */
export const parseDocument = (payload) => new JSDOM(payload).window.document
