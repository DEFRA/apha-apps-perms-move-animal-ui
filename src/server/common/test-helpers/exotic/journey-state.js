export const aboutSectionVisitComplete = {
  movementType: 'visit'
}

export const aboutSectionNonVisitComplete = {
  movementType: 'onto-premises',
  whatIsMoving: 'carcasses',
  whatAreYouMovingDetails: 'pig carcasses',
  howMuchAreYouMoving: '10 pig carcasses',
  whatAreYouMovingDescription: 'culled pig carcasses for incineration'
}
