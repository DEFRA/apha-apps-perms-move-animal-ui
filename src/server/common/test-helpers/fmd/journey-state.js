export const aboutSectionComplete = {
  movementActivityType: 'slaughter-onsite',
  animalSlaughtered: 'cattle',
  slaughteredNumber: 123,
  animalIds: 'ear tags'
}
