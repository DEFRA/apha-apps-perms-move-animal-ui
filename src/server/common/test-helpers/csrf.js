import { statusCodes } from '~/src/server/common/constants/status-codes.js'
/* global expect, test */

/**
 * @param {OutgoingHttpHeaders} headers
 * @param {string} name
 * @returns {string}
 */
const findCookie = (headers, name) => {
  const setCookie = headers['set-cookie']

  /** @type {string[] | undefined} */
  const cookies = typeof setCookie === 'string' ? [setCookie] : setCookie
  return cookies?.find((h) => h?.includes(`${name}=`)) ?? ''
}

/**
 * @param {string} cookieString
 */
const extractCookieValue = (cookieString) =>
  cookieString.split('=')[1].split(';')[0]

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedPost = (serverFn, injectionOptions) =>
  test('should reject if experiencing an apparent csrf attack (they have their session cookie, but no hidden form value)', async () => {
    const { statusCode } = await serverFn().inject({
      ...injectionOptions,
      headers: {
        ...(injectionOptions.headers ?? {}),
        Cookie: 'crumb=csrf-value'
      }
    })

    expect(statusCode).toEqual(statusCodes.forbidden)
  })

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedGet = (serverFn, injectionOptions) =>
  test('should contain a cookie with a set CSRF value, and contain a hidden form field to submit that CSRF back to the server', async () => {
    const { headers, payload } = await serverFn().inject(injectionOptions)

    const token = extractCookieValue(findCookie(headers, 'crumb'))
    expect(token).not.toHaveLength(0)
    expect(payload).toContain(
      `<input type="hidden" name="crumb" value="${token}"`
    )
  })

/**
 * @param {ServerInjectOptions} injectionOptions
 * @returns {ServerInjectOptions}
 */
export const withCsrfProtection = (injectionOptions) => ({
  ...injectionOptions,
  payload: {
    ...(typeof injectionOptions.payload === 'object'
      ? injectionOptions.payload
      : {}),
    crumb: 'csrf-value'
  },
  headers: {
    ...(injectionOptions.headers ?? {}),
    Cookie: 'crumb=csrf-value'
  }
})

/**
 * @import { OutgoingHttpHeaders } from 'http'
 * @import { ServerInjectOptions, Server } from '@hapi/hapi'
 */
