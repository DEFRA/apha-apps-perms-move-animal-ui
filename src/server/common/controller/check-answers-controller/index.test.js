import { createServer } from '~/src/server/index.js'
import { statusCodes } from '~/src/server/common/constants/status-codes.js'
import { withCsrfProtection } from '~/src/server/common/test-helpers/csrf.js'
import { parseDocument } from '~/src/server/common/test-helpers/dom.js'
import SessionTestHelper from '../../test-helpers/session-helper.js'
import { validApplicationState } from '../../test-helpers/journey-state.js'
import { spyOnConfig } from '../../test-helpers/config.js'
import { sizeErrorPage } from '../../../tb/biosecurity-map/size-error/index.js'

import Wreck from '@hapi/wreck'
import Boom from '@hapi/boom'
import { config } from '~/src/config/config.js'

/**
 * @import { IncomingMessage } from 'http'
 */

const {
  origin,
  destination,
  licence,
  identification,
  biosecurity,
  'biosecurity-map': biosecurityMap
} = validApplicationState
const pageTitle = 'Check your answers before sending your application'
const confirmationUri = '/tb/submit/confirmation'
const checkAnswersUri = '/tb/submit/check-answers'

describe('#CheckAnswers', () => {
  /** @type {Server} */
  let server
  let session

  beforeAll(async () => {
    server = await createServer()
    await server.initialize()
  })

  beforeEach(async () => {
    session = await SessionTestHelper.create(server)
    await session.setSectionState('origin', origin)
    await session.setSectionState('licence', licence)
    await session.setSectionState('destination', destination)
    await session.setSectionState('identification', identification)
    await session.setSectionState('biosecurity', biosecurity)
    await session.setSectionState('biosecurity-map', biosecurityMap)
  })

  afterAll(async () => {
    await server.stop({ timeout: 0 })
  })

  it('Should provide expected response', async () => {
    const { payload, statusCode } = await server.inject(
      withCsrfProtection(
        {
          method: 'GET',
          url: checkAnswersUri,
          payload: {}
        },
        {
          Cookie: session.sessionID
        }
      )
    )

    const document = parseDocument(payload)
    expect(document.querySelector('#main-content')?.innerHTML).toMatchSnapshot()
    expect(document.title).toEqual(pageTitle)
    expect(statusCode).toBe(statusCodes.ok)
  })

  it('Should stay in check-answers if all tasks are valid', async () => {
    const { statusCode } = await server.inject(
      withCsrfProtection(
        {
          method: 'GET',
          url: checkAnswersUri
        },
        {
          Cookie: session.sessionID
        }
      )
    )

    expect(statusCode).toBe(statusCodes.ok)
  })

  describe('when sendToCaseManagement feature flag is enabled', () => {
    it('should return a 500 error if submit fails for any other reason', async () => {
      const wreckSpy = jest.spyOn(Wreck, 'post').mockResolvedValue({
        res: /** @type {IncomingMessage} */ ({
          statusCode: 400
        }),
        payload: JSON.stringify({
          message: 'An error occured'
        })
      })

      const { statusCode } = await server.inject(
        withCsrfProtection(
          {
            method: 'POST',
            url: checkAnswersUri + '?case-management-api=true',
            payload: {
              confirmation: 'confirm'
            }
          },
          {
            Cookie: session.sessionID
          }
        )
      )

      expect(statusCode).toBe(statusCodes.serverError)
      expect(wreckSpy).toHaveBeenCalledTimes(1)
    })

    it('should redirect to the sizze error page if the file is too large', async () => {
      const wreckSpy = jest.spyOn(Wreck, 'post').mockImplementation(() => {
        const error = Boom.badRequest('Dummy error')
        error.output.statusCode = statusCodes.fileTooLarge
        throw error
      })

      const { headers, statusCode } = await server.inject(
        withCsrfProtection(
          {
            method: 'POST',
            url: checkAnswersUri + '?case-management-api=true',
            payload: {
              confirmation: 'confirm'
            }
          },
          {
            Cookie: session.sessionID
          }
        )
      )

      expect(statusCode).toBe(statusCodes.redirect)
      expect(headers.location).toBe(sizeErrorPage.urlPath)
      expect(wreckSpy).toHaveBeenCalledTimes(1)
    })

    it('should send the application to case management when feature flag is disabled but query string present', async () => {
      const dummyReferenceNumber = '12-1234-1234'
      spyOnConfig('featureFlags', {
        sendToCaseManagement: false
      })

      const wreckSpy = jest.spyOn(Wreck, 'post').mockResolvedValue({
        res: /** @type {IncomingMessage} */ ({
          statusCode: 200
        }),
        payload: JSON.stringify({
          message: dummyReferenceNumber
        })
      })

      const { headers, statusCode } = await server.inject(
        withCsrfProtection(
          {
            method: 'POST',
            url: checkAnswersUri + '?case-management-api=true',
            payload: {
              confirmation: 'confirm'
            }
          },
          {
            Cookie: session.sessionID
          }
        )
      )

      expect(statusCode).toBe(statusCodes.redirect)
      expect(headers.location).toBe(confirmationUri)

      const { reference, 'state-key': stateKey } = await session.getState(
        'tb-confirmation-details'
      )
      expect(reference).toBe(dummyReferenceNumber)
      expect(stateKey).toBe('application')
      expect(wreckSpy).toHaveBeenCalledTimes(1)
      expect(wreckSpy.mock.calls[0][0]).toBe(
        `${config.get('caseManagementApi').baseUrl}/submit`
      )
    })

    it('should give the user a 500 error if the case management API doesnt work for what ever reason', async () => {
      spyOnConfig('featureFlags', {
        sendToCaseManagement: true
      })

      jest.spyOn(Wreck, 'post').mockResolvedValue({
        res: /** @type {IncomingMessage} */ ({
          statusCode: 502
        }),
        payload: ''
      })

      const { statusCode } = await server.inject(
        withCsrfProtection(
          {
            method: 'POST',
            url: checkAnswersUri,
            payload: {
              confirmation: 'confirm'
            }
          },
          {
            Cookie: session.sessionID
          }
        )
      )

      expect(statusCode).toBe(statusCodes.serverError)
    })

    it('should send the application to case management and redirect to confirmation page', async () => {
      const dummyReferenceNumber = '12-1234-1234'
      spyOnConfig('featureFlags', {
        sendToCaseManagement: true
      })

      const wreckSpy = jest.spyOn(Wreck, 'post').mockResolvedValue({
        res: /** @type {IncomingMessage} */ ({
          statusCode: 200
        }),
        payload: JSON.stringify({
          message: dummyReferenceNumber
        })
      })

      const { headers, statusCode } = await server.inject(
        withCsrfProtection(
          {
            method: 'POST',
            url: checkAnswersUri,
            payload: {
              confirmation: 'confirm'
            }
          },
          {
            Cookie: session.sessionID
          }
        )
      )

      expect(statusCode).toBe(statusCodes.redirect)
      expect(headers.location).toBe(confirmationUri)

      const { reference } = await session.getState('tb-confirmation-details')

      expect(reference).toBe(dummyReferenceNumber)
      expect(wreckSpy).toHaveBeenCalledTimes(1)
      expect(wreckSpy.mock.calls[0][0]).toBe(
        `${config.get('caseManagementApi').baseUrl}/submit`
      )
    })
  })
})

/**
 * @import { Server } from '@hapi/hapi'
 */
