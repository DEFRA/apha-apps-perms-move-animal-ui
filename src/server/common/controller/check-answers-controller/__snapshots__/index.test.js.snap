// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#CheckAnswers Should provide expected response 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <h1 class="govuk-heading-l">Check your answers before sending your application</h1>
    </div>
  </div>
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <div class="govuk-body" data-testid="app-page-body">
          <h2 class="govuk-heading-m">Movement origin</h2>
          
  <dl class="govuk-summary-list govuk-!-margin-bottom-9">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Are you moving the animals on or off your farm or premises?
    </dt>
    <dd class="govuk-summary-list__value">
      On to the farm or premises
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/origin/to-or-from-own-premises?redirect_uri=/tb/submit/check-answers" data-testid="onOffFarm-change-link">Change<span class="govuk-visually-hidden"> Are you moving the animals on or off your farm or premises?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which type of premises are the animals moving off?
    </dt>
    <dd class="govuk-summary-list__value">
      TB restricted farm
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/origin/type-of-origin?redirect_uri=/tb/submit/check-answers" data-testid="originType-change-link">Change<span class="govuk-visually-hidden"> Which type of premises are the animals moving off?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What is the county parish holding (CPH) number of the farm or premises where the animals are moving off?
    </dt>
    <dd class="govuk-summary-list__value">
      12/345/6789
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/origin/origin-farm-cph?redirect_uri=/tb/submit/check-answers" data-testid="cphNumber-change-link">Change<span class="govuk-visually-hidden"> What is the county parish holding (CPH) number of the farm or premises where the animals are moving off?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What is the address of the farm or premises where the animals are moving off?
    </dt>
    <dd class="govuk-summary-list__value">
      Starfleet Headquarters<br>San Francisco<br>RG24 8RR
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/origin/origin-farm-address?redirect_uri=/tb/submit/check-answers" data-testid="address-change-link">Change<span class="govuk-visually-hidden"> What is the address of the farm or premises where the animals are moving off?</span></a>
    </dd>
  </div>
</dl>

          <h2 class="govuk-heading-m">Movement destination</h2>
          
  <dl class="govuk-summary-list govuk-!-margin-bottom-9">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Where are the animals going to?
    </dt>
    <dd class="govuk-summary-list__value">
      TB restricted farm
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/destination/type-of-destination?redirect_uri=/tb/submit/check-answers" data-testid="destinationType-change-link">Change<span class="govuk-visually-hidden"> Where are the animals going to?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What is the county parish holding (CPH) number of the farm or premises where the animals are going to?
    </dt>
    <dd class="govuk-summary-list__value">
      12/345/6789
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/destination/destination-farm-cph?redirect_uri=/tb/submit/check-answers" data-testid="destinationFarmCph-change-link">Change<span class="govuk-visually-hidden"> What is the county parish holding (CPH) number of the farm or premises where the animals are going to?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What is the address of the farm or premises where the animals are going to?
    </dt>
    <dd class="govuk-summary-list__value">
      Starfleet Headquarters<br>San Francisco<br>RG24 8RR
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/destination/destination-farm-address?redirect_uri=/tb/submit/check-answers" data-testid="destinationFarmAddress-change-link">Change<span class="govuk-visually-hidden"> What is the address of the farm or premises where the animals are going to?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      How many animals are you planning to move?
    </dt>
    <dd class="govuk-summary-list__value">
      1
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/destination/how-many-animals?redirect_uri=/tb/submit/check-answers" data-testid="howManyAnimals-change-link">Change<span class="govuk-visually-hidden"> How many animals are you planning to move?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What is the reason for the movement?
    </dt>
    <dd class="govuk-summary-list__value">
      Welfare
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/destination/reason-for-movement?redirect_uri=/tb/submit/check-answers" data-testid="reasonForMovement-change-link">Change<span class="govuk-visually-hidden"> What is the reason for the movement?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Enter any additional information (optional)
    </dt>
    <dd class="govuk-summary-list__value">
      
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/destination/any-additional-info?redirect_uri=/tb/submit/check-answers" data-testid="additionalInfo-change-link">Change<span class="govuk-visually-hidden"> Enter any additional information (optional)</span></a>
    </dd>
  </div>
</dl>

          <h2 class="govuk-heading-m">Receiving the licence</h2>
          
  <dl class="govuk-summary-list govuk-!-margin-bottom-9">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Who will be the registered owner of the animals?
    </dt>
    <dd class="govuk-summary-list__value">
      Kathryn Janeway
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/receiving-the-licence/licence-name-future?redirect_uri=/tb/submit/check-answers" data-testid="fullName-change-link">Change<span class="govuk-visually-hidden"> Who will be the registered owner of the animals?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What email address would you like the licence sent to?
    </dt>
    <dd class="govuk-summary-list__value">
      kathryn@starfleet.com
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/receiving-the-licence/licence-enter-email-address?redirect_uri=/tb/submit/check-answers" data-testid="emailAddress-change-link">Change<span class="govuk-visually-hidden"> What email address would you like the licence sent to?</span></a>
    </dd>
  </div>
</dl>

          <h2 class="govuk-heading-m">Animal identification</h2>
          
  <dl class="govuk-summary-list govuk-!-margin-bottom-9">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Will you move any calves under 42 days old?
    </dt>
    <dd class="govuk-summary-list__value">
      No
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/identification/any-calves?redirect_uri=/tb/submit/check-answers" data-testid="calvesUnder42DaysOld-change-link">Change<span class="govuk-visually-hidden"> Will you move any calves under 42 days old?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What are the dates of the last TB tests for the animals that are 42 days old or older?
    </dt>
    <dd class="govuk-summary-list__value">
      01/01/2025
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/identification/enter-testing-dates?redirect_uri=/tb/submit/check-answers" data-testid="testingDates-change-link">Change<span class="govuk-visually-hidden"> What are the dates of the last TB tests for the animals that are 42 days old or older?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Enter the ear tag numbers for these animals
    </dt>
    <dd class="govuk-summary-list__value">
      123456
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/identification/enter-ear-tags-over-42-days?redirect_uri=/tb/submit/check-answers" data-testid="earTags-change-link">Change<span class="govuk-visually-hidden"> Enter the ear tag numbers for these animals</span></a>
    </dd>
  </div>
</dl>

          <h2 class="govuk-heading-m">Biosecurity details</h2>
          
  <dl class="govuk-summary-list govuk-!-margin-bottom-9">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Will you separate the incoming animals from the resident herd?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/kept-separately?redirect_uri=/tb/submit/check-answers" data-testid="keptSeparately-change-link">Change<span class="govuk-visually-hidden"> Will you separate the incoming animals from the resident herd?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Will the incoming animals be grazed?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/grazing?redirect_uri=/tb/submit/check-answers" data-testid="grazing-change-link">Change<span class="govuk-visually-hidden"> Will the incoming animals be grazed?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      How long ago was the field last grazed by the resident herd?
    </dt>
    <dd class="govuk-summary-list__value">
      yesterday
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/last-grazed?redirect_uri=/tb/submit/check-answers" data-testid="lastGrazed-change-link">Change<span class="govuk-visually-hidden"> How long ago was the field last grazed by the resident herd?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Has any manure or slurry been put on the grazing field in the past 60 days?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/manure-and-slurry?redirect_uri=/tb/submit/check-answers" data-testid="manureAndSlurry-change-link">Change<span class="govuk-visually-hidden"> Has any manure or slurry been put on the grazing field in the past 60 days?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which measures are being taken to reduce the spread of TB when the animals are grazing?
    </dt>
    <dd class="govuk-summary-list__value">
      A minimum of 3 metres will separate the animals
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/grazing-field-how-separated?redirect_uri=/tb/submit/check-answers" data-testid="grazingFieldHowSeparated-change-link">Change<span class="govuk-visually-hidden"> Which measures are being taken to reduce the spread of TB when the animals are grazing?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      How will you manage slurry and manure?
    </dt>
    <dd class="govuk-summary-list__value">
      Manure and slurry is securely stored so domestic and wild animals can't access it
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/manure-and-slurry-details?redirect_uri=/tb/submit/check-answers" data-testid="manureAndSlurryDetails-change-link">Change<span class="govuk-visually-hidden"> How will you manage slurry and manure?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      What disinfectant are you using?
    </dt>
    <dd class="govuk-summary-list__value">
      some disinfectant
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/disinfectant?redirect_uri=/tb/submit/check-answers" data-testid="disinfectant-change-link">Change<span class="govuk-visually-hidden"> What disinfectant are you using?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Confirmation of the dilution rate
    </dt>
    <dd class="govuk-summary-list__value">
      I confirm a dilution rate of 1:15 is used on the farm or premises
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/disinfectant-dilution?redirect_uri=/tb/submit/check-answers" data-testid="dilutionRate-change-link">Change<span class="govuk-visually-hidden"> Confirmation of the dilution rate</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Will the incoming animals share any equipment and machinery with the resident herd?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/equipment-any-shared?redirect_uri=/tb/submit/check-answers" data-testid="equipmentShared-change-link">Change<span class="govuk-visually-hidden"> Will the incoming animals share any equipment and machinery with the resident herd?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which measures are in place to clean and disinfect equipment to reduce the risk of spreading TB?
    </dt>
    <dd class="govuk-summary-list__value">
      Designated disinfection points
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/equipment-how-minimise-contamination?redirect_uri=/tb/submit/check-answers" data-testid="equipmentHowMinimiseContamination-change-link">Change<span class="govuk-visually-hidden"> Which measures are in place to clean and disinfect equipment to reduce the risk of spreading TB?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Will the incoming animals be housed?
    </dt>
    <dd class="govuk-summary-list__value">
      Yes
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/buildings-any-shared?redirect_uri=/tb/submit/check-answers" data-testid="animalsHoused-change-link">Change<span class="govuk-visually-hidden"> Will the incoming animals be housed?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      How will you reduce the risk of spreading TB from the resident herd to the incoming animals during housing?
    </dt>
    <dd class="govuk-summary-list__value">
      somehow
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/buildings-how-minimise-contamination?redirect_uri=/tb/submit/check-answers" data-testid="buildingsHowMinimiseContamination-change-link">Change<span class="govuk-visually-hidden"> How will you reduce the risk of spreading TB from the resident herd to the incoming animals during housing?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which measures are staff taking to reduce the risk of spreading TB?
    </dt>
    <dd class="govuk-summary-list__value">
      Dedicated clothing and personal protective equipment (PPE)
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/people-disinfection?redirect_uri=/tb/submit/check-answers" data-testid="peopleDisinfection-change-link">Change<span class="govuk-visually-hidden"> Which measures are staff taking to reduce the risk of spreading TB?</span></a>
    </dd>
  </div>
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Which measures are you taking to reduce the risk of infection from wildlife?
    </dt>
    <dd class="govuk-summary-list__value">
      Badger-proof fencing, such as wired fences, solid aluminium sheeted gates and rail fences or retractable electric fences
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity/badgers?redirect_uri=/tb/submit/check-answers" data-testid="badgers-change-link">Change<span class="govuk-visually-hidden"> Which measures are you taking to reduce the risk of infection from wildlife?</span></a>
    </dd>
  </div>
</dl>

          <h2 class="govuk-heading-m">Biosecurity map</h2>
          
  <dl class="govuk-summary-list govuk-!-margin-bottom-9">
  <div class="govuk-summary-list__row">
    <dt class="govuk-summary-list__key govuk-!-width-one-half govuk-!-font-weight-regular">
      Upload a biosecurity map
    </dt>
    <dd class="govuk-summary-list__value">
      Map uploaded
    </dd>
    <dd class="govuk-summary-list__actions">
      <a class="govuk-link" href="/biosecurity-map/upload-plan?redirect_uri=/tb/submit/check-answers" data-testid="upload-plan-change-link">Change<span class="govuk-visually-hidden"> Upload a biosecurity map</span></a>
    </dd>
  </div>
</dl>

      </div>
    </div>
  </div>
  <h2 class="govuk-heading-m">Your declaration</h2>
    <form method="POST" novalidate="">
      <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds">
          <input type="hidden" name="crumb" value="csrf-value">
          <input type="hidden" name="nextPage" value="">

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset">
    <legend class="govuk-fieldset__legend">
      Before you submit your application, you must select one of the declaration check boxes.
    </legend>
    <div class="govuk-checkboxes" data-module="govuk-checkboxes">
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input" id="confirmation" name="confirmation" type="checkbox" value="confirm" data-testid="confirm-checkbox">
        <label class="govuk-label govuk-checkboxes__label" for="confirmation">
          I confirm all the information given about my farm or premises is correct to the best of my knowledge
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input" id="confirmation-2" name="confirmation" type="checkbox" value="other" data-testid="other-checkbox">
        <label class="govuk-label govuk-checkboxes__label" for="confirmation-2">
          I am submitting this form on behalf of someone else and confirm all the information given is correct to the best of my knowledge
        </label>
      </div>
    </div>
  </fieldset>
</div>

            <button type="submit" class="govuk-button" data-module="govuk-button" id="continue-button">
  Continue
</button>
        </div>
      </div>
    </form>
        "
`;
